### PR TITLE
Simplistic support for reporting lexing and parsing errors in template files

### PR DESCRIPTION
**Pull request by gasche**
This PR is intended to make progress towards fixing #42 : indeed, if we want to use this for human-written templates, we should support the cases where the input is invalid and provide meaningful error messages.

The current PR does not implement meaningful error messages, but at least it documents the possibility of failure and gives the position where the error was detected by the lexer or parser.

It is not completely clear to me how much error-explanation cleverness you want in the main library. Arguably this would rather be the domain of "real applications" using the library down the chain, and the library could limit itself to giving information about the lexing/parsing state on error, to be formatted nicely by library users. This PR uses a different approach where the error type is abstract, and formatting is done by the library; this is more convenient in the short term, but less flexible in the long term.

Note: if we wanted good parsing error messages, we could not just use Menhir's support as-is, because currently most of the work is actually done by the lexer which "bundles" delimiters together with their payloads. So a first step would be to de-bundle the lexer, by instead emitting a token for the delimiter and one or several tokens for the payload, giving more grammar information to Menhir. (We could still keep some context-dependent lexing by having an explicit representation of the lexing state in the lexer: are we in-between delimiters?, etc.) This would be nice, but a much more invasive change than what I had the courage to do in a first PR.

### CODE REVIEW

**Comment by rgrinberg** [comment]
> It is not completely clear to me how much error-explanation cleverness you want in the main library.

Probably as much as possible. The consumers of such a templating library are likely using it to template html or simple configuration files. It would be great to give them a good experience out of the box. If this isn't flexible enough for some consumers, we can always add a more complex api on top.

> if we wanted good parsing error messages, we could not just use Menhir's support as-is, because currently most of the work is actually done by the lexer which "bundles" delimiters together with their payloads. So a first step would be to de-bundle the lexer, by instead emitting a token for the delimiter and one or several tokens for the payload, giving more grammar information to Menhir

Sounds good to me.

File: lib/mustache.ml [hunk 1]
```diff
@@ -142,14 +142,55 @@ let to_string x =
   Buffer.contents b
 
 (* Parsing: produces an ast with locations. *)
+type template_parse_error = {
+  lexbuf: Lexing.lexbuf;
+  kind: error_kind;
+}
+and error_kind = Parsing | Lexing of string
+
+exception Parse_error of template_parse_error
```
Comment (rgrinberg) [inline lib/mustache.ml:151]: Should we register a printer for this exception?

Comment (gasche) [reply lib/mustache.ml:151]: Good point, I added exception-registration logic in a new commit.

**Comment by gasche** [comment]
Now that I have slightly more experience with the `ocaml-mustache` codebase, I see several things with this PR that are not quite right, and I think should be fixed.
- the `Invalid_template` error is also a parse-time error and should be handled as such.
- The `pp_error` name is too generic, as we probably also want better exception payloads for render-time ("runtime") errors, which would need a separate error type.

I hope to fix them shortly so that users don't see the intermediary state. I am planning to propose other improvements to error information and error reporting to the library, but those would change existing exceptions so they would break some compatibility -- unlike this PR.

**Pull request merged**